NuRadioReco/utilities/noise.py

```diff
@@ -100,3 +203,391 @@ def generate_noise(self):
+        """ separated trigger part for PA noise trigger using np.lib.stride_tricks.as_strided """
+        self.max_amp = 0
+        for iBeam, phased_trace in enumerate(self._phased_traces):
+            # Create a sliding window
+            coh_sum_squared = phased_trace ** 2
+            num_frames = int(np.floor((len(phased_trace) - self.window) / self.step))
+            coh_sum_windowed = np.lib.stride_tricks.as_strided(coh_sum_squared, (num_frames, self.window),
+                                                       (coh_sum_squared.strides[0] * self.step, coh_sum_squared.strides[0]))
+            squared_mean = np.sum(coh_sum_windowed, axis=1) / self.window
+            #self.max_amp = max(squared_mean.max(), self.max_amp)
+            self.max_amp = max(squared_mean.max(), self.max_amp)
+            if True in (squared_mean > self.threshold):
+                print(f"triggered at beam {iBeam}")
+                if(self.debug):
+                    import matplotlib.pyplot as plt
+                    fig, ax = plt.subplots(5, 1, sharex=True)
+                    for iCh in range(self.n_channels):
+                        ax[iCh].plot(self._traces[iCh])
+                        print(f"{self._traces[iCh].std():.2f}")
+                    ax[4].plot(self._phased_traces[iBeam])
+                    fig.tight_layout()
+                    plt.show()
+                return True
+        return False
+
+    def generate_noise_optimised(self, debug=False):
+        """
+        generates noise traces for all channels that will cause a high/low majority logic trigger
+
+        Returns np.array of shape (n_channels, n_samples)
+        """
+        self.debug = debug
+        self.max_amp = 0
+        # some variables for profiling code
+        dt_generation = 0
+        dt_phasing = 0
+        dt_triggering = 0
+
+        # generate empty trace array
+        self._traces = np.zeros((self.n_channels, self.n_samples * self.upsampling))
+        counter = 0
+        max_amp = 0
+        #print("noise trace length:", self.n_samples * self.upsampling)
+        while True:
+            counter += 1
+            if(counter % 1000 == 0):
+                print(f"{counter:d}, {self.max_amp:.2f}, threshold = {self.threshold:.2f}")
+                # some printout for profiling
+                print(f"GENERATION: {dt_generation:.4f}, PHASING: {dt_phasing:.4f}, TRIGGER: {dt_triggering:.4f}")
+            tstart = time.process_time()
+            self.__generation()
+            dt_generation += time.process_time() - tstart
+
+            tstart = time.process_time()
+            self.__phasing()
+            dt_phasing += time.process_time() - tstart
+
+            tstart = time.process_time()
+            is_triggered = self.__triggering()
+            dt_triggering += time.process_time() - tstart
+            if is_triggered:
+                return self._traces, self._phased_traces
+
+
+    def generate_noise2(self, debug=False):
```

Comment:
why is this function in here? Maybe a left over from the other noise class? If so, then just remove this function.